master/buildbot/www/auth.py

```diff
@@ -131,6 +136,36 @@ def maybeAutoLogin(self, request):
             yield self.updateUserInfo(request)
 
 
+class LDAPAuth(AuthBase):
+    """
+    Authenticate user against LDAP
+    """
+    def __init__(self, server_uri: str, binddn: Optional[str] = None, bindpw: Optional[str] = None, **kwargs):
+        """
+        :param server_uri: Server connection string (example: "ldaps://ldap.company.com:636")
+        :param binddn: The user permitted to search the LDAP directory
+        :param bindpw: Password for binddn user
+        """
+        super().__init__(**kwargs)
+        self.server_uri = server_uri
+        self.binddn = binddn
+        self.bindpw = bindpw
+
+    def requestAvatarId(self, cred):
```

Comment:
This is what I have so far:

```python
    def authenticate(self, cred):
        server = urlparse(self.server_uri)
        s = ldap3.Server(server.hostname, port=server.port, use_ssl=server.scheme == 'ldaps', get_info=ldap3.ALL)
        with ldap3.Connection(s, auto_bind=True, client_strategy=ldap3.SYNC,
                              user=self.bb_bind_user, password=self.bb_bind_password,
                              authentication=ldap3.SIMPLE) as c:
            # Make sure we can connect to the server with the bind credentials
            if c.bind():
                # Search for the user
                res = yield c.search(self.account_subtree, search_pattern.format(username=cred.username))

                if not res or len(res.entries) == 1:
                    # Check if the credentials can authenticate
                    if c.rebind(user=cred.username, password=cred.password):
                        return defer.succeed(cred.username)
            else:
                return defer.fail(LDAPBindError())
        return defer.fail(UnauthorizedLogin())

    def requestAvatarId(self, cred):
        return threads.deferToThread(self.authenticate, cred)
```

In your code, you create a closure and then defer it to a thread. In my code, I defer the bound `self.authenticate` method, including its `cred` argument, to a thread.

So, a few questions:

1. Is there a difference between our two approaches? Or are they basically the same and it's just a stylistic choice?
2. Which method do I decorate with `defer.inlineCallbacks`? `requestAvatarId` or `authenticate`? Or both?
3. The `renderSecrets` method is only available on `buildbot.www.service.BuildbotService`, does that mean that `LDAPAuth` needs to inherit from `BuildbotService`? Or should I reimplement that method elsewhere? Or does it make sense to move the `renderSecrets` method to a different mixin class? I don't think it relies on anything else in `BuildbotService`, so it might make more sense to refactor it to a different mixin.

Reply:
Sorry for not noticing this comment. To answer your questions:

1. There's effectively no difference. Your approach is fine.
2. @defer.inlineCallbacks is used when you want to "wait" for deferreds using yield. In your above code, if `c.search` returns a Deferred, then you would write `res = yield c.search(...)` and decorate the function with @defer.inlineCallbacks. If you get a Deferred and you don't do anything after that, it's fine just to return it. The following are equivalent, though the non-inlineCallbacks case is faster.
```
def foo(self, arg1):
   return self.function_that_returns_a_deferred(arg1)

@defer.inlineCallbacks
def foo(self, arg1):
    res = yield self.function_that_returns_a_deferred(arg1)
    return res
```

Several gotchas:

twisted documentation says one needs to use defer.returnValue to return values from @inlineCallbacks functions. This is only needed in python2 code, we can use regular returns.

The current implementation of `authenticate` does not seem to actually use any deferred operations. Looking into the old code I _think_ that c.search does not return a Deferred, but I may be wrong.

It looks that there's no reason for defer.succeed and defer.fail. defer.succeed should be replaced by a regular return and defer.fail should be replaced by throwing an exception.

3. Yes, you'll need to inherit from `BuildbotService`. It contains additional special code for secrets in `reconfigServiceWithSibling`. In theory I agree that `renderSecrets` could go to a different mixin, but in the end that mixin would always be used together with `BuildbotService` so it's not very useful to split it away.